doc/compiler/control/CheckpointRestoreCoordination.md

```diff
@@ -0,0 +1,434 @@
+suspend itself for Checkpoint. This inconsistency can only create a deadlock if
+the Checkpointing Thread tries to acquire the IProfiler Monitor with the CR
+Monitor in hand (which it never does).
+
+The coordination of the IProfiler Thread is a little different from the Sampler
+Thread because the unlike the Sampler Thread which only performs a timed sleep,
+the IProfiler Thread waits until it needs to process a buffer that is added to
+the working queue. Additionally, the IProfiler will continue to process buffers
+before it suspends itself so that it does not have any outstanding buffers post
+restore.
+
+### Shutdown
+
+The Shutdown process can be initiated at any point above where the Comp Monitor
+or the IProfiler Monitor is not held by either the Checkpointing or the
+IProfiler Thread. Therefore, in Step 4, the Checkpointing Thread also checks if
+the checkpoint should be interrupted, and breaks returns if so. On the side of
+the IProfiler Thread, shutdown cannot occur once the IProfiler Thread acquires
+the Comp Monitor and verifies that the Checkpoint Status is
+`TR_CheckpointStatus::SUSPEND_THREADS_FOR_CHECKPOINT`. If Shutdown occurs prior
+to that, the IProfiler Thread will not execute the main body of
+`suspendIProfilerThreadForCheckpoint` and terminate normally; if Shutdown occurs
+after the IProfiler Thread has suspended itself, the Shutdown Thread will notify
+the IProfiler Thread to initiate shutdown via the IProfiler Thread Lifetime
+State.
+
+# Preparing for Restore
+
+As mentioned above, `TR::CompilationInfo::prepareForRestore` handles all the
+coordination needed to prepare the compiler for restore.
+
+First this method calls `J9::OptionsPostRestore::processOptionsPostRestore` to
+process any options set post-restore. Once this task is done, it acquires the
+Comp Monitor. It then resets the Checkpoint Status. Next it resets the Start
+time, and proceeds to invoke `resumeJITThreadsForRestore`, which
+1. Resumes the IProfiler Thread
+2. Resumes the Sampler Thread
+3. Resumes the Compilation Threads
+
+It then ensures that Swap Memory is availble before returning back to the VM.
+
+## Resetting the Start Time
+While the Checkpoint phase is conceptually part of building the application, in
+order to ensure consistency with parts of the compiler that memoize elapsd time,
+the start time is reset to pretend like the JVM started
+`persistentInfo->getElapsedTime()` milliseconds prior to time at restore.
+
+## Resuming the IProfiler Thread
+The Checkpointing Thread at this point has the Comp Monitor in hand.
+1. The Checkpointing Thread acquires the IProfiler Monitor
+2. The Checkpointing Thread sets the IProfiler Thread Lifetime State to
+   `TR_IProfiler::IPROF_THR_RESUMING`
+3. The Checkpointing Thread calls `notifyAll`, and releases the IProfiler
+   Monitor
+4. The IProfiler Thread resumes with the IProfiler Monitor in hand; it releases
+   the IProfiler Monitor, acquires the Comp Monitor, and then reacquires the
```

Comment:
Yeah it will, the IProfiler will remain blocked until the Checkpointing thread releases the Comp Monitor (possibly after finishing up the restore process). However, this is ok because we're resuming, and so the IProfiler Thread will eventually get its hand on the monitor and resume.